### PR TITLE
feat: accept DatasetListItem where DatasetReference is accepted

### LINKED ISSUE #588: feat: `list_tables` accepts a `DatasetListItem`
**Is your feature request related to a problem? Please describe.**

Code:

```python
for dataset in bqclient.list_datasets():
    tables = list(bqclient.list_tables(dataset, max_results=1))
    if len(tables) == 0:
        bqclient.delete_dataset(dataset)
```

Output:

```
Traceback (most recent call last):
  File "2021/04-test-dataset-cleanup/cleanup.py", line 11, in <module>
    tables = len(bqclient.list_tables(dataset, max_results=1))
  File "/Users/swast/miniconda3/envs/scratch/lib/python3.8/site-packages/google/cloud/bigquery/client.py", line 1327, in list_tables
    raise TypeError("dataset must be a Dataset, DatasetReference, or string")
TypeError: dataset must be a Dataset, DatasetReference, or string
```

 **Describe the solution you'd like**

Where `Dataset` or `DatasetReference` is accepted, also accept a `DatasetListItem`.

 **Describe alternatives you've considered**

I know that we created `DatasetListItem.reference`, which I can use, but it'd be a bit more intuitive if I didn't have to.

 **Additional context**

N/A

### PR DESCRIPTION
**Pull request by jimfulton**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #588 🦕

### CODE REVIEW

**Review by plamut — COMMENTED** [review]
Looking good so far. :+1:

**Review by plamut — APPROVED** [review]
Nice cleanup work, thanks!

File: tests/unit/test_create_dataset.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
+from google.cloud.bigquery.dataset import Dataset, DatasetReference
```
Comment (tswast) [inline tests/unit/test_create_dataset.py:1]: Need to have Apache Header on all code files

https://opensource.google/docs/releasing/preparing/#Apache-header

Comment (jimfulton) [reply tests/unit/test_create_dataset.py:1]: On it. :)

**Pull request merged**